tools/storage.py

```diff
@@ -41,29 +43,144 @@ def __init__(self, batch):
+        v = v + key_length
+
+        (value_length, v) = read_varlong(b, v)
+        value = b[v:v + value_length]
+        v = v + value_length
+
+        (header_count, v) = read_varlong(b, v)
+        headers = []
+        for i in range(0, header_count):
+            (header_length, v) = read_varlong(b, v)
+            header = b[v:v + header_length]
+            v = v + header_length
+            headers.append(header)
+
+        return Record(size, record_attr, timestamp_delta, offset_delta, key,
+                      headers, value)
+
+    @staticmethod
+    def from_bytes(b, count):
+        records = []
+        for i in range(0, count):
+            record = Record.record_from_bytes(b)
+            b = b[:record.size]
+            records.append(record)
+        return records
+
+
 class Batch:
     def __init__(self, index, header, records):
         self.index = index
         self.header = header
+        self.records = records
 
+    def verify_crc(self):
         header_crc_bytes = struct.pack(
             "<" + HDR_FMT_RP_PREFIX_NO_CRC + HDR_FMT_CRC, *self.header[1:])
-        header_crc = crc32c.crc32(header_crc_bytes)
+        header_crc = crc32c.crc32c(header_crc_bytes)
         if self.header.header_crc != header_crc:
             raise CorruptBatchError(self)
 
-        crc = crc32c.crc32(self._crc_header_be_bytes())
-        crc = crc32c.crc32(records, crc)
+        crc = crc32c.crc32c(self._crc_header_be_bytes())
+        crc = crc32c.crc32c(self.records, crc)
         if self.header.crc != crc:
             raise CorruptBatchError(self)
 
     def last_offset(self):
-        return self.header.base_offset + self.header.record_count - 1
+        return self.header.base_offset + self.header.delta
 
     def _crc_header_be_bytes(self):
         # encode header back to big-endian for crc calculation
         return struct.pack(">" + HDR_FMT_CRC, *self.header[5:])
 
+    def parse_records(self):
+        batch = self.records
+        try:
+            if self.header.attrs == 0:
+                pass
+            elif self.header.attrs == 1:
+                import gzip
+                logger.warning("Found zip")
+                batch = gzip.decompress(self.records)
+            elif self.header.attrs == 2:
+                "snappy"
+                raise NotImplementedError(
+                    "Unsupported compression type snappy. Patches welcome!")
+            elif self.header.attrs == 3:
+                import lz4.frame
+                logger.warning("Found lz4")
+                batch = lz4.frame.decompress(self.records)
+            elif self.header.attrs == 4:
+                import zstd
+                batch = zstd.decompress(self.records)
+            else:
+                raise NotImplementedError(
+                    f"Unknown compression type {self.header.attrs}. Patches welcome!"
+                )
+
+            return Record.from_bytes(batch, self.header.record_count)
+        except zstd.Error as e:
```

Comment:
if we are already testing for `zstd` then we should allow the user to percolate the error, it means data corruption tho.

Reply:
if you want to allow zstd corruption, maybe we have a flag that says `--skip-corrupted` that swallows the exception like we do here?

Reply:
actually come to think of it, this script is not ready to handled corrupted data, since the CRC will throw

so we don't need exception handling here since it fails on the crc check *before* it gets to check for compression. so we can remove this exception handling

Reply:
I *think* you're right, now that it properly detects the compression and validates crc it should always work. I've removed the error handling and will re-run against my big corrupted blob and we'll see.

Reply:
I'm leaving the error handling in locally so the job finishes, but no need to push the change. It's up to you, really.

Reply:
Ok so I am definitely getting batches that are marked as zstd, pass crc checks, and then somehow fail to decompress using zstd. Any ideas what could cause such a scenario? Lots of messages like this at the top of my run:

```
WARNING:storage:Unable to decode batch, Input data invalid or missing content size in frame header. Header(header_crc=4046115462, batch_size=46356, base_offset=371315879, type=1, crc=3270988256, attrs=4, delta=704, first_ts=1619227515435, max_ts=1619227544386, producer_id=-1, producer_epoch=-1, base_seq=-1, record_count=705) from b'(\xb5/\xfd\x00X<(\x02\xce\xc7\x04g\x0bB\xc0\xda\x00`H\x960\r\xc9B\xe2\x96iH\x1e\x82\xe4'
```

Secondly, I fixed a bug causing rec.count copies of record 1 instead of reading each record due to using the wrong python slicing operator. 

Chasing this down, I also found a bug in decoding headers. Instead of reading headers as a dict, I was reading them as strings and missing  half of them. Headers now read properly (into a dict). 

Finally, in the process of tracking this down, I determined that the rec.size read from the header is two bytes short of the offset needed to start reading the next batch (storage.py:125).

Is this expected?